frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,381 @@ pub mod pallet {
+		/// Merge two vesting schedules together, creating a new vesting schedule that unlocks over the
+		/// highest possible start and end blocks. If both schedules have already started the current
+		/// block will be used as the schedule start; with the caveat that if one schedule is finished
+		/// by the current block, the other will be treated as the new merged schedule, unmodified.
+		///
+		/// NOTE: If `schedule1_index == schedule2_index` this is a no-op.
+		/// NOTE: This will unlock all schedules through the current block prior to merging.
+		/// NOTE: If both schedules have ended by the current block, no new schedule will be created
+		/// and both will be removed.
+		///
+		/// Merged schedule attributes:
+		/// - `starting_block`: `MAX(schedule1.starting_block, scheduled2.starting_block, current_block)`.
+		/// - `ending_block`: `MAX(schedule1.ending_block, schedule2.ending_block)`.
+		/// - `locked`: `schedule1.locked_at(current_block) + schedule2.locked_at(current_block)`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schedule1_index`: index of the first schedule to merge.
+		/// - `schedule2_index`: index of the second schedule to merge.
+		#[pallet::weight(
+			T::WeightInfo::not_unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get())
+			.max(T::WeightInfo::unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get()))
+		)]
+		pub fn merge_schedules(
+			origin: OriginFor<T>,
+			schedule1_index: u32,
+			schedule2_index: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			if schedule1_index == schedule2_index {
+				return Ok(());
+			};
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+			let schedules = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+
+			// The schedule index is based off of the schedule ordering prior to filtering out any
+			// schedules that may be ending at this block.
+			let schedule1 =
+				*schedules.get(schedule1_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let schedule2 =
+				*schedules.get(schedule2_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let merge_action = VestingAction::Merge(schedule1_index, schedule2_index);
+
+			// The length of `schedules` decreases by 2 here since we filter out 2 schedules.
+			// Thus we know below that we can insert the new merged schedule without error (assuming
+			// initial state was valid).
+			let (mut schedules, mut locked_now) =
+				Self::report_schedule_updates(schedules.to_vec(), merge_action);
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			let now = <frame_system::Pallet<T>>::block_number();
+			if let Some(new_schedule) = Self::merge_vesting_info(now, schedule1, schedule2)? {
```

Comment:
This there a reason this logic isnt also inside the `report_schedule_updates`

The action is `Merge`, and yet there is some follow up action that is needed here to make it work.

I would expect the function to simply return what values should be saved to storage.

Reply:
I would prefer to leave `report_schedule_updates` alone so it does not get too complicated. But I think it is fair to wrap that in another function for merging.

But maybe this `VestingAction` name is a misnomer, its really just a fixed length filter at this level of abstraction.

Reply:
yeah vesting action does not represent what is going on here.

My thought was that you wanted some function which you gave it the values and the thing you wanted to do, and it spit out the final values you need to commit to storage.

If you want to design it differently, def you need to update the naming/apis here.

Reply:
I essentially implemented your design prescription here: 4d410f7 . Check out `exec_action` for the meat of the logic